File: src/ast/BaseNode.java
package ast;

import java.io.StringWriter;

public abstract class BaseNode {
    public NodeType Type;

    public BaseNode(NodeType type) {
        Type = type;
    }

    //virtual
    public void Print(StringWriter writer) {
        PrintLeft(writer);

        if (HasRightPart()) {
            PrintRight(writer);
        }
    }

    public abstract void PrintLeft(StringWriter writer);

    //virtual
    public boolean HasRightPart() {
        return false;
    }

    //virtual
    public boolean IsArray() {
        return false;
    }

    //virtual
    public boolean HasFunctions() {
        return false;
    }

    //virtual
    public String GetName() {
        return null;
    }

    //virtual
    public void PrintRight(StringWriter writer) {
    }

    @Override
    public String toString() {
        StringWriter writer = new StringWriter();

        Print(writer);

        return writer.toString();
    }
}

File: src/de/fearlesstobi/demangler/ast/BaseNode.kt
package de.fearlesstobi.demangler.ast

import java.io.StringWriter

abstract class BaseNode internal constructor(var type: NodeType) {
    //virtual
    fun print(writer: StringWriter) {
        printLeft(writer)
        if (hasRightPart()) {
            printRight(writer)
        }
    }

    abstract fun printLeft(writer: StringWriter)
    //virtual
    open fun hasRightPart(): Boolean {
        return false
    }

    //virtual
    open val isArray: Boolean
        get() = false

    //virtual
    open fun hasFunctions(): Boolean {
        return false
    }

    //virtual
    open fun getName(): String?{
        return ""
    }

    //virtual
    open fun printRight(writer: StringWriter) {}

    override fun toString(): String {
        val writer = StringWriter()
        print(writer)
        return writer.toString()
    }

}